src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnectionImpl.java

```diff
@@ -1411,16 +1427,38 @@ private Object doPrivilegedOperation(final int operation,
         serverCommunicatorAdmin.reqIncoming();
         try {
             PrivilegedOperation op = new PrivilegedOperation(operation, params);
-            if (acc == null) {
-                try {
-                    return op.run();
-                } catch (Exception e) {
-                    if (e instanceof RuntimeException)
-                        throw (RuntimeException) e;
-                    throw new PrivilegedActionException(e);
+            if (!SharedSecrets.getJavaLangAccess().allowSecurityManager()) {
+                // Modern case
+                if (subject == null) {
+                    try {
+                        return op.run();
+                    } catch (Exception e) {
+                        if (e instanceof RuntimeException)
+                           throw (RuntimeException) e;
+                        throw new PrivilegedActionException(e);
+                    }
+                } else {
+                    return Subject.doAs(subject, op);
                 }
             } else {
-                return AccessController.doPrivileged(op, acc);
+                // SM permitted
+                if (acc == null) {
+                    try {
+                        return op.run();
+                    } catch (Exception e) {
+                        if (e instanceof RuntimeException)
+                           throw (RuntimeException) e;
+                        throw new PrivilegedActionException(e);
+                    }
+                } else {
+                    return AccessController.doPrivileged(op, acc);
+                }
+            }
+        } catch (Exception e) {
+            if (e instanceof RuntimeException rte) {
+                throw rte;
+            } else {
+                throw new PrivilegedActionException(e);
```

Comment:
The 4 lines above seems unnecessary now. Plus, do not wrap `PrivilegedActionException` inside `PrivilegedActionException`.

Reply:
It needs to recognise and throw RuntimeException so that a SecurityException isn't wrapped in a PrivilegedActionException (which gets caught by those blocks of code which call extractException(pe) and  look at what Exception it contains).

(test/jdk/javax/management/remote/mandatory/notif/NotificationAccessControllerTest.java tests this)

I can add a check to make sure PrivilgedActionException doesn't get wrapped.  Daniel I think mentioned this also.

Reply:
I still don't understand how a SecurityException could be wrapped into a PAE without these lines.

Reply:
I misunderstood the comment, yes I think I can remove this block now.